src/celeritas/ext/detail/GeantMicroXsCalculator.cc

```diff
@@ -0,0 +1,114 @@
+//----------------------------------*-C++-*----------------------------------//
+// Copyright 2023 UT-Battelle, LLC, and other Celeritas developers.
+// See the top-level COPYRIGHT file for details.
+// SPDX-License-Identifier: (Apache-2.0 OR MIT)
+//---------------------------------------------------------------------------//
+//! \file celeritas/ext/detail/GeantMicroXsCalculator.cc
+//---------------------------------------------------------------------------//
+#include "GeantMicroXsCalculator.hh"
+
+#include <CLHEP/Units/SystemOfUnits.h>
+#include <G4Material.hh>
+#include <G4VEmModel.hh>
+
+#include "corecel/cont/Range.hh"
+
+namespace celeritas
+{
+namespace detail
+{
+//---------------------------------------------------------------------------//
+/*!
+ * Construct with Geant4 classes.
+ *
+ * The secondary production cut should be nonzero if a secondary is produced.
+ */
+GeantMicroXsCalculator::GeantMicroXsCalculator(
+    G4VEmModel const& model,
+    G4ParticleDefinition const& particle,
+    G4Material const& material,
+    double secondary_production_cut)
+    : model_{const_cast<G4VEmModel&>(model)}
+    , particle_{particle}
+    , material_{material}
+    , secondary_cut_{secondary_production_cut}
+{
+    CELER_EXPECT(secondary_cut_ >= 0);
+}
+
+//---------------------------------------------------------------------------//
+/*!
+ * Calculate cross sections for all elements in the material.
+ */
+void GeantMicroXsCalculator::operator()(VecDouble const& energy_grid,
+                                        VecVecDouble* result_xs) const
+{
+    CELER_EXPECT(result_xs);
+
+    std::vector<G4Element const*> const& elements
+        = *material_.GetElementVector();
+
+    // Resize microscopic cross sections for all elements
+    result_xs->resize(elements.size());
+    for (auto& mxs_vec : *result_xs)
+    {
+        mxs_vec.resize(energy_grid.size());
+    }
+
+    // Outer loop over energy to reduce material setup calls
+    for (auto energy_idx : range(energy_grid.size()))
+    {
+        double const energy = energy_grid[energy_idx];
+        model_.SetupForMaterial(&particle_, &material_, energy);
+
+        // Inner loop over elements
+        for (auto elcomp_idx : range(elements.size()))
+        {
+            (*result_xs)[elcomp_idx][energy_idx]
+                = this->calc_element_xs(*elements[elcomp_idx], energy);
+        }
+    }
+
+    for (std::vector<double>& xs : *result_xs)
+    {
+        // Avoid cross-section vectors starting or ending with zero values.
+        // Geant4 simply uses the next/previous bin value when the vector's
+        // front/back are zero. This probably isn't correct but it replicates
+        // Geant4's behavior.
```

Comment:
Why not? I guess it's just to make sure the xs is constant outside the grid.

Reply:
My thought was that there's no reason the second energy grid point will physically be the first nonzero cross section; so the cross section is actually changing at some point between the two grid points. Leaving the first value as zero means incorrectly linearly interpolating; copying the first behavior means a possibly incorrect extrapolation.

This was implemented in #326 and I remember @stognini explaining his "this feels like a trap" comment that was in there. I'm actually not sure now that our implementation is correct: if Geant4 extrapolates the lowest xs value below the xs range, but replaces the in-bounds zero xs value with the lowest nonzero xs value, then it *will* have zero below the energy bounds and nonzero inside, whereas we'll be nonzero everywhere...

Reply:
Looks like the `G4PhysicsVector` just returns the first value if the energy is below the grid like we do, so it should be constant and nonzero outside.

Reply:
Another point is that `G4VEmModel::ComputeCrossSectionPerAtom()` could still return `xs[0] = xs[1] = 0` from a given model implementation...

Reply:
Not totally sure, but I think in general as long as the energy is above `LowEnergyLimit` (or cutoff energy) the computed cross section will be positive?

Reply:
🤷‍♂️ I would hope so? I don't recall precisely where in Geant4 the zero-cross section replacement is done...